### PR TITLE
feat: generate gif from chat directly

### LINKED ISSUE #2: [feat] Generate GIF directly from RC Chat
## Details

The feature would include several steps:
- [ ] Implement a Slash Command, which in-turn sends an HTTP request to our API Endpoint.
- [ ] Implement two API Endpoints, one of which will be used as an API dispatcher which will send a request to replicate. The other endpoint will be passed as a WebHook to the Replicate API.
- [ ] Implement Persistence to store the status of the generations done by the user. Store the requestID,  queuedMessageID, roomID, userID etc.
- [ ] After receiving the GIF through WebHook, get the relevant data stored in the persistence from the provided `id` and populate the old message with the generated GIF.

### PR DESCRIPTION
**Pull request by AmanNegi**
## Tasks:

- [x] Implement Slash Command to take prompt from the user.
- [x] Setup preferences for several properties like (`API_KEY, API_URL, WEBHOOK_URL, MODEL_ID`).
- [x] Validate, preferences are as per requirements before attempting GIF generation.
- [x] Create a `RequestDispatcher` which is responsible for interacting with the remote API's. (The API's can be easily changed from the preferences, till the response format is sustained).
- [x] Implement, APIEndpoint/Webhook which can be notified once the generation has been completed.
- [x] Create a Persistence storage to store the ongoing generations and the details regarding them, so that they can be fetched when we receive an update via webhooks.

> Fixes: #2 

## Demo

[feat-generate-gif.webm](https://github.com/RocketChat/Apps.AI.GIF/assets/37607224/c0fc048a-a16a-4c52-9360-92d4a7a65305)

### CODE REVIEW

**Comment by AmanNegi** [comment]
@Nabhag8848 Could you please take a look at this PR. Let me know if any changes are required.

File: definition/lib/IGifResponseData.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export enum PredictionStatus {
+    Starting = "starting",
```
Comment (Nabhag8848) [inline definition/lib/IGifResponseData.ts]: @AmanNegi its good practice to always write enum keys in capital: 

```
enum PrediectionStatus{
      STARTING = "starting",
      ...similarly others. 
}

```
also you might change there is a typo in Cancel status.

File: src/commands/GenGifCommand.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    constructor(private readonly app: AiGifApp) {}
+
+    async executor(
+        context: SlashCommandContext,
+        read: IRead,
+        modify: IModify,
+        http: IHttp,
+        persis: IPersistence
+    ): Promise<void> {
+        const dispatcher = new GifRequestDispatcher(
```
Comment (Nabhag8848) [inline src/commands/GenGifCommand.ts:32]: Can you make a CommandUtility Class to manage different params command, see reference below. which will help us. to manage different commands later.
https://github.com/RocketChat/Apps.Notion/blob/main/src/commands/CommandUtility.ts

File: src/lib/GifRequestDispatcher.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            input: {
+                mp4: false,
+                steps: 30,
+                width: 672,
+                height: 384,
+                prompt,
+                negative_prompt: "blurry",
+            },
+        };
+
```
Comment (Nabhag8848) [inline src/lib/GifRequestDispatcher.ts:127]: Need to make this settings avialable to users at later point.

**Review by Nabhag8848 — CHANGES_REQUESTED** [review]
@AmanNegi Added few comments lemme know once you fix those.

File: definition/lib/IGifResponseData.ts (hunk 1 above)
Comment (AmanNegi) [reply definition/lib/IGifResponseData.ts]: Okay, I'll make those changes. Regarding the typo this is an issue in the [Replicate API](https://replicate.com/docs/reference/http#predictions.get), where cancelled is spelled as canceled. I'll fix the key to match our requirements if that's fine.

Comment (Nabhag8848) [reply definition/lib/IGifResponseData.ts]: Cool, for typo add a comment over there.

File: src/persistence/OnGoingGenPersistence.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                {
+                    records: [
+                        record,
+                        ...(records[0] as OnGoingGenRecordWrapper).records,
+                    ],
+                }
+            );
+        }
+    }
+}
```
Comment (Nabhag8848) [inline src/persistence/OnGoingGenPersistence.ts:102]: - Make this association with user, as discussed @AmanNegi

File: src/utils/prefs.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+): Promise<string | undefined> {
+    const settingValue = await read
+        .getEnvironmentReader()
+        .getSettings()
+        .getValueById(id);
+
+    if (!settingValue || settingValue === "") {
+        return undefined;
+    }
+
```
Comment (Nabhag8848) [inline src/utils/prefs.ts:15]:
```suggestion

```
This will fail if we don't pass URL, as discussed.

File: src/commands/GenGifCommand.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import { AiGifApp } from "../../AiGifApp";
+import { sendMessageToRoom, sendMessageToSelf } from "../utils/message";
+import { InfoMessages } from "../enum/InfoMessages";
+import { GifRequestDispatcher } from "../lib/GifRequestDispatcher";
+import { OnGoingGenPersistence } from "../persistence/OnGoingGenPersistence";
+
+export class GenGifCommand implements ISlashCommand {
+    command = "gen-gif";
+    i18nParamsExample = "gen-gif your-prompt-here";
+    i18nDescription = "Generate a gif based on your prompt";
```
Comment (Nabhag8848) [inline src/commands/GenGifCommand.ts]: - Implement i18n as discussed in correct way

**Review by Nabhag8848 — CHANGES_REQUESTED** [review]

File: src/lib/GifRequestDispatcher.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    async mockGenerateGif(
+        prompt: string,
+        id: string
+    ): Promise<IGifResponseData> {
+        const webhookUrl = await getSettingFromId(
+            this.read,
+            Preferences.WEBHOOK_URL
+        );
+
+        console.log("Mocking generation of gif", webhookUrl);
```
Comment (Nabhag8848) [inline src/lib/GifRequestDispatcher.ts]: Can you remove this console.log, never push console.log

File: src/commands/GenGifCommand.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+            `${InfoMessages.GENERATION_IN_PROGRESS}\nPrompt: ${prompt}`,
+            ":hourglass_flowing_sand:"
+        );
+
+        const onGoingGenPeristence = new OnGoingGenPersistence(
+            persis,
+            read.getPersistenceReader()
+        );
+
+        console.log("ICommand user id: ", context.getSender().id);
```
Comment (Nabhag8848) [inline src/commands/GenGifCommand.ts]:
```suggestion
```
this too

**Review by Nabhag8848 — CHANGES_REQUESTED** [review]
Hey @AmanNegi can you remove this console logs, lemme know once this done. we can merge this.

File: src/lib/GifRequestDispatcher.ts (hunk 7 above)
Comment (AmanNegi) [reply src/lib/GifRequestDispatcher.ts]: Sorry missed this out! fixing this real quick

**Review by Nabhag8848 — APPROVED** [review]
LGTM @AmanNegi Merging this.

**Pull request merged**